scripts/calculate_rating.py

```diff
@@ -0,0 +1,151 @@
+'''
+Implementation of following rating algorithm
+https://codeforces.com/blog/entry/20762
+'''
+
+import datetime
+import pymysql
+import json
+from pymysql.cursors import DictCursor
+
+import itertools
+import pprint
+
+USER_QUERY = 'SELECT user_id, username from users'
+
+# TODO technically if a user has the same start time, they'll show up twice, this could be an issue
+RUNS_QUERY = ('''
+SELECT runs.run_id, runs.prompt_id, runs.user_id, users.username,
+        TIMESTAMPDIFF(MICROSECOND, runs.start_time, runs.end_time) AS run_time
+FROM runs 
+RIGHT JOIN users ON runs.user_id=users.user_id 
+INNER JOIN (
+    SELECT user_id, MIN(start_time) as start_time
+    FROM runs
+    GROUP BY runs.prompt_id, runs.user_id
+) t ON runs.user_id = t.user_id AND runs.start_time = t.start_time
+ORDER BY runs.prompt_id, run_time;
+''')
+
+STORE_RATINGS_QUERY = (
+'''
+REPLACE INTO ratings (user_id, rating, num_rounds) VALUES (%s, %s, %s);
+'''
+)
+
+INITIAL_RATING = 1500
+
+pp = pprint.PrettyPrinter(indent=4)
+
+def _elo_prob(ri, rj):
+    return 1 / (1 + 10 ** ((rj - ri) / 400 ))
+
+def _calculate_seed(users, round):
+    for u in round:
+        users[u]["seed"] = 1
+        for v in round:
+            if (u == v): continue
+            users[u]["seed"] += _elo_prob(users[v]["rating"], users[u]["rating"])
+
+
+def _calculate_place(users, round):
+    for place, u in enumerate(round):
+        users[u]["place"] = (place + 1)
+
+def _calculate_desired_seed(users):
+    for u in users:
+        user = users[u]
+        if ("seed" not in user or "place" not in user): continue
+
+        user["desired_seed"] = (user["seed"] * user["place"]) ** 0.5
+
+def _rating_for_seed(users, round, u, desired_seed):
+    lo = 1
+    hi = 8000
+
+    while (lo < hi):
+        mid = (lo + hi) // 2
+
+        mid_seed = 1
+
+        # TODO should we exclude ourselves?
```

Comment:
I think the person setting the prompt of the day probably shouldn't play it - honor system is fine lol

Reply:
Oh this is something with the rating system, whether your own rating factors into calculating your expected place